Error while checking build/O2/fullCI for 2325014f445bdcca9608f719641f32869807e337 at 2024-04-03 16:51:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/12980-slc8_x86-64/0/Generators/src/BoxGenerator.cxx:48:24: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12980/2325014f445bdcca9608f719641f32869807e337/build_O2_fullCI/pretty.html).

--- a/Generators/CMakeLists.txt
+++ b/Generators/CMakeLists.txt
@@ -32,6 +32,7 @@ o2_add_library(Generators
                        src/TriggerExternalParam.cxx
                        src/TriggerParticleParam.cxx
                        src/BoxGunParam.cxx
+                       src/BoxGenerator.cxx
                        src/QEDGenParam.cxx
                        src/GenCosmicsParam.cxx
                        src/GeneratorFactory.cxx
@@ -80,6 +81,7 @@ set(headers
     include/Generators/GeneratorTParticle.h
     include/Generators/GeneratorTParticleParam.h
     include/Generators/GeneratorService.h
+    include/Generators/BoxGenerator.h
     )
 
 if(pythia_FOUND)

--- a/Generators/include/Generators/BoxGenerator.h
+++ b/Generators/include/Generators/BoxGenerator.h
@@ -1,0 +1,113 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \author Sandro Wenzel - April 2024
+
+#ifndef ALICEO2_EVENTGEN_BOX
+#define ALICEO2_EVENTGEN_BOX
+
+#include "Generators/Generator.h"
+#include "TParticle.h"
+#include <vector>
+
+namespace o2::eventgen
+{
+
+/*
+ * A simple mono-pdg "BoxGenerator". Or particle gun.
+ * Re-implements FairBoxGenerator for more convenient O2-processing.
+ */
+class BoxGenerator : public Generator
+{
+ public:
+  BoxGenerator() = default;
+  BoxGenerator(int pdgid, int mult = 1);
+
+  BoxGenerator(int pdgid,
+               int mult,
+               double etamin,
+               double etamax,
+               double pmin,
+               double pmax,
+               double phimin,
+               double phimax) : mPDG{pdgid}, mMult{mult}
+  {
+    SetEtaRange(etamin, etamax);
+    SetPRange(pmin, pmax);
+    SetPhiRange(phimin, phimax);
+  }
+
+  void SetPRange(Double32_t pmin = 0, Double32_t pmax = 10)
+  {
+    mPMin = pmin;
+    mPMax = pmax;
+    mPRangeIsSet = true;
+  }
+
+  void SetPhiRange(double phimin = 0, double phimax = 360)
+  {
+    mPhiMin = phimin;
+    mPhiMax = phimax;
+  }
+
+  void SetEtaRange(double etamin = -5, double etamax = 5)
+  {
+    mEtaMin = etamin;
+    mEtaMax = etamax;
+    mEtaRangeIsSet = true;
+  }
+
+  /// generates a single particle conforming to particle gun parameters
+  TParticle sampleParticle() const;
+
+  /// implements the main O2 generator interfaces
+  bool generateEvent() override
+  {
+    mEvent.clear();
+    for (int i = 0; i < mMult; ++i) {
+      mEvent.push_back(sampleParticle());
+    }
+    return true;
+  }
+  bool importParticles() override
+  {
+    mParticles.clear();
+    std::copy(mEvent.begin(), mEvent.end(), std::back_insert_iterator(mParticles));
+    return true;
+  }
+
+ private:
+  double mPtMin{0.}, mPtMax{0.};       // Transverse momentum range [GeV]
+  double mPhiMin{0.}, mPhiMax{360.};   // Azimuth angle range [degree]
+  double mEtaMin{0.}, mEtaMax{0.};     // Pseudorapidity range in lab system
+  double mYMin{0.}, mYMax{0.};         // Rapidity range in lab system
+  double mPMin{0.}, mPMax{0.};         // Momentum range in lab system
+  double mThetaMin{0.}, mThetaMax{0.}; // Polar angle range in lab system [degree]
+  double mEkinMin{0.}, mEkinMax{0.};   // Kinetic Energy range in lab system [GeV]
+
+  int mPDG{0};
+  int mMult{1};
+
+  bool mEtaRangeIsSet{false};   // True if eta range is set
+  bool mYRangeIsSet{false};     // True if rapidity range is set
+  bool mThetaRangeIsSet{false}; // True if theta range is set
+  bool mCosThetaIsSet{false};   // True if uniform distribution in
+  // cos(theta) is set (default -> not set)
+  bool mPtRangeIsSet{false};   // True if transverse momentum range is set
+  bool mPRangeIsSet{false};    // True if abs.momentum range is set
+  bool mEkinRangeIsSet{false}; // True if kinetic energy range is set
+
+  std::vector<TParticle> mEvent; // internal event container
+};
+
+} // namespace o2::eventgen
+
+#endif

--- a/Generators/include/Generators/Generator.h
+++ b/Generators/include/Generators/Generator.h
@@ -71,8 +71,8 @@ class Generator : public FairGenerator
   Bool_t ReadEvent(FairPrimaryGenerator* primGen) final;
 
   /** methods to override **/
-  virtual Bool_t generateEvent() = 0;
-  virtual Bool_t importParticles() = 0;
+  virtual Bool_t generateEvent() = 0;   // generates event (in structure internal to generator)
+  virtual Bool_t importParticles() = 0; // fills the mParticles vector (transfer from generator state)
 
   /** setters **/
   void setMomentumUnit(double val) { mMomentumUnit = val; };

--- a/Generators/src/BoxGenerator.cxx
+++ b/Generators/src/BoxGenerator.cxx
@@ -1,0 +1,78 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \author Sandro Wenzel - April 2024
+
+#include "Generators/BoxGenerator.h"
+#include "TRandom.h"
+#include "TDatabasePDG.h"
+
+using namespace o2::eventgen;
+
+double GetPDGMass(int pdg)
+{
+  static TDatabasePDG* pid = TDatabasePDG::Instance();
+  TParticlePDG* p = pid->GetParticle(pdg);
+  if (p != nullptr) {
+    // LOG(info) << this->ClassName() << ": particle with PDG =" << GetPDGType() << " Found";
+    return p->Mass(); // fPDGMass = p->Mass();
+  }
+  // LOG(warn) << "pdg not known";
+  return 0.;
+}
+
+TParticle o2::eventgen::BoxGenerator::sampleParticle() const
+{
+  // Primary particles are distributed uniformly along
+  // those kinematics variables which were limitted by setters.
+  // if SetCosTheta() function is used, the distribution will be uniform in
+  // cos(theta)
+
+  static double mass = GetPDGMass(mPDG);
+
+  double pabs = 0, phi, pt = 0, theta = 0, eta, y, mt, px, py, pz = 0;
+  phi = gRandom->Uniform(mPhiMin, mPhiMax) * TMath::DegToRad();
+  if (mPRangeIsSet) {
+    pabs = gRandom->Uniform(mPMin, mPMax);
+  } else if (mPtRangeIsSet) {
+    pt = gRandom->Uniform(mPtMin, mPtMax);
+  }
+  if (mThetaRangeIsSet) {
+    if (mCosThetaIsSet) {
+      theta = acos(gRandom->Uniform(cos(mThetaMin * TMath::DegToRad()), cos(mThetaMax * TMath::DegToRad())));
+    } else {
+      theta = gRandom->Uniform(mThetaMin, mThetaMax) * TMath::DegToRad();
+    }
+  } else if (mEtaRangeIsSet) {
+    eta = gRandom->Uniform(mEtaMin, mEtaMax);
+    theta = 2 * TMath::ATan(TMath::Exp(-eta));
+  } else if (mYRangeIsSet) {
+    y = gRandom->Uniform(mYMin, mYMax);
+    mt = TMath::Sqrt(mass * mass + pt * pt);
+    pz = mt * TMath::SinH(y);
+  }
+
+  if (mThetaRangeIsSet || mEtaRangeIsSet) {
+    if (mPRangeIsSet) {
+      pz = pabs * TMath::Cos(theta);
+      pt = pabs * TMath::Sin(theta);
+    } else if (mPtRangeIsSet) {
+      pz = pt / TMath::Tan(theta);
+    }
+  }
+  px = pt * TMath::Cos(phi);
+  py = pt * TMath::Sin(phi);
+
+  double vx = 0., vy = 0., vz = 0.;
+  double etot = TMath::Sqrt(px * px + py * py + pz * pz + mass * mass);
+  return TParticle(mPDG, 1 /*status*/, -1 /* mother1 */, -1 /* mother2 */,
+                   -1 /* daughter1 */, -1 /* daughter2 */, px, py, pz, etot, vx, vy, vz, 0. /*time*/);
+}

--- a/Generators/src/Generator.cxx
+++ b/Generators/src/Generator.cxx
@@ -75,19 +75,23 @@ Bool_t
 
     /** generate event **/
     if (!generateEvent()) {
+      LOG(error) << "ReadEvent failed in generateEvent";
       return kFALSE;
     }
 
     /** import particles **/
     if (!importParticles()) {
+      LOG(error) << "ReadEvent failed in importParticles";
       return kFALSE;
     }
 
     if (mSubGeneratorsIdToDesc.empty() && mSubGeneratorId > -1) {
+      LOG(error) << "ReadEvent failed because no SubGenerator description given";
       return kFALSE;
     }
 
     if (!mSubGeneratorsIdToDesc.empty() && mSubGeneratorId < 0) {
+      LOG(error) << "ReadEvent failed because SubGenerator description given but sub-generator not set";
       return kFALSE;
     }
 
@@ -102,6 +106,7 @@ Bool_t
 
   /** add tracks **/
   if (!addTracks(primGen)) {
+    LOG(error) << "ReadEvent failed in addTracks";
     return kFALSE;
   }
 

--- a/Generators/src/GeneratorsLinkDef.h
+++ b/Generators/src/GeneratorsLinkDef.h
@@ -73,4 +73,6 @@
 #pragma link C++ class o2::eventgen::GeneratorFileOrCmdParam + ;
 #pragma link C++ class o2::conf::ConfigurableParamHelper < o2::eventgen::GeneratorFileOrCmdParam> + ;
 
+#pragma link C++ class o2::eventgen::BoxGenerator + ;
+
 #endif

